velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+    auto searchIndex = searchIndices[row];
+
+    auto instance = instanceDecoded.valueAt<int64_t>(row);
+    VELOX_USER_CHECK_NE(
+        instance,
+        0,
+        "array_position cannot take a 0-valued instance argument.");
+
+    int startIndex = instance > 0 ? 0 : size - 1;
+    int endIndex = instance > 0 ? size : -1;
+    int step = instance > 0 ? 1 : -1;
+    instance = std::abs(instance);
+
+    for (auto i = startIndex; i != endIndex; i += step) {
+      if (!elementsBase->isNullAt(offset + i) &&
+          elementsBase->equalValueAt(searchBase, offset + i, searchIndex)) {
+        --instance;
+        if (instance == 0) {
+          flatResult.set(row, i + 1);
+          return;
+        }
+      }
+    }
+    flatResult.set(row, 0);
+  });
+}
+
+template <>
+void applyTypedWithInstance<TypeKind::ARRAY>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    const DecodedVector& instanceDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeWithInstance(
+      rows,
+      arrayDecoded,
+      elementsDecoded,
+      searchDecoded,
+      instanceDecoded,
+      flatResult);
+}
+
+template <>
+void applyTypedWithInstance<TypeKind::MAP>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    const DecodedVector& instanceDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeWithInstance(
+      rows,
+      arrayDecoded,
+      elementsDecoded,
+      searchDecoded,
+      instanceDecoded,
+      flatResult);
+}
+
+template <>
+void applyTypedWithInstance<TypeKind::ROW>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    const DecodedVector& instanceDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeWithInstance(
+      rows,
+      arrayDecoded,
+      elementsDecoded,
+      searchDecoded,
+      instanceDecoded,
+      flatResult);
+}
+
+class ArrayPositionFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    const auto& arrayVector = args[0];
+    const auto& searchVector = args[1];
+    VELOX_CHECK(arrayVector->type()->isArray());
+    VELOX_CHECK(arrayVector->type()->asArray().elementType()->kindEquals(
+        searchVector->type()));
+
+    BaseVector::ensureWritable(rows, BIGINT(), context->pool(), result);
+    auto flatResult = (*result)->asFlatVector<int64_t>();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto elements = decodedArgs.at(0)->base()->as<ArrayVector>()->elements();
+    exec::LocalSelectivityVector nestedRows(context, elements->size());
```

Comment:
@mbasmanova I am not sure this is correct; But it would be great if you can validate or correct me here.

As per my understanding we shouldnt try to access all elements here but only those required by rows - i.e we should use the toElementRows function instead of assuming that we need to decode all elements.

Reply:
@kgpai That's a good question. This code works. See https://github.com/facebookincubator/velox/pull/501 for more context.

Reply:
Good question. I actually followed the code in `ArrayContains.cpp`, but it's great to understand the reason behind it.